crates/optimism/node/src/engine.rs

```diff
@@ -146,7 +146,100 @@ where
+                ))
+            }
+        }
 
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod test {
+
+    use alloy_primitives::{Address, B256};
+    use alloy_rpc_types_engine::PayloadAttributes;
+    use reth_chainspec::ForkCondition;
+    use reth_optimism_chainspec::BASE_SEPOLIA;
+    use crate::engine;
+
+    use super::*;
+
+    fn pre_holocene_chainspec() -> Arc<OpChainSpec> {
+        let hardforks = OptimismHardfork::base_sepolia();
+        Arc::new(OpChainSpec {
+            inner: ChainSpec {
+                chain: BASE_SEPOLIA.inner.chain,
+                genesis: BASE_SEPOLIA.inner.genesis.clone(),
+                genesis_hash: BASE_SEPOLIA.inner.genesis_hash.clone(),
+                paris_block_and_final_difficulty: BASE_SEPOLIA.inner.paris_block_and_final_difficulty,
+                hardforks,
+                base_fee_params: BASE_SEPOLIA.inner.base_fee_params.clone(),
+                max_gas_limit: BASE_SEPOLIA.inner.max_gas_limit,
+                prune_delete_limit: 10000,
+                ..Default::default()
+            },
+        })
+    }
+
+
+    fn holocene_chainspec() -> Arc<OpChainSpec> {
+        let mut hardforks = OptimismHardfork::base_sepolia();
+        hardforks.insert(OptimismHardfork::Holocene.boxed(), ForkCondition::Timestamp(1800000000));
+        Arc::new(OpChainSpec {
+            inner: ChainSpec {
+                chain: BASE_SEPOLIA.inner.chain,
+                genesis: BASE_SEPOLIA.inner.genesis.clone(),
+                genesis_hash: BASE_SEPOLIA.inner.genesis_hash.clone(),
+                paris_block_and_final_difficulty: BASE_SEPOLIA.inner.paris_block_and_final_difficulty,
+                hardforks,
+                base_fee_params: BASE_SEPOLIA.inner.base_fee_params.clone(),
+                max_gas_limit: BASE_SEPOLIA.inner.max_gas_limit,
+                prune_delete_limit: 10000,
+                ..Default::default()
+            },
+        })
+    }
+
+    #[test]
+    fn test_well_formed_attributes_pre_holocene() {
+        let validator = OptimismEngineValidator::new(pre_holocene_chainspec());
+
+        let attributes = OpPayloadAttributes {
+            gas_limit: Some(1000),
+            eip_1559_params: None,
+            transactions: None,
+            no_tx_pool: None,
+            payload_attributes: PayloadAttributes {
+                timestamp: 0x1337,
+                prev_randao: B256::ZERO,
+                suggested_fee_recipient: Address::ZERO,
+                withdrawals: Default::default(),
+                parent_beacon_block_root: Some(B256::ZERO),
+            },
+        };
+
+        // assert!(<engine::OptimismEngineValidator as reth_node_builder::EngineValidator<EngineTypes<PayloadAttributes = OpPayloadAttributes>>>::ensure_well_formed_attributes(&validator, EngineApiMessageVersion::V4, &attributes).is_ok());
+        // assert!(validator.ensure_well_formed_attributes(EngineApiMessageVersion::V4, &attributes).is_ok());
```

Comment:
So I'm trying to write unit tests for the invalid attribute but it's quite tricky.

This line just doesn't seem to work, it returns a bunch of type errors. No matter how I change it the syntax/types just don't work. Any suggestions?

Reply:
@clabby @refcell @Rjected wondering if you guys have any suggestions on this

Reply:
you have to specify all of the engine types, rather than just `= OpPayloadAttributes`, for example

```rust
        assert!(<engine::OptimismEngineValidator as reth_node_builder::EngineValidator<OptimismEngineTypes>>::ensure_well_formed_attributes(&validator, EngineApiMessageVersion::V4, &attributes).is_ok());
```

Reply:
@Rjected Nice thanks! Fixed